clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -502,6 +518,83 @@ synchronized void updateCache(boolean onDemandRefresh) throws InterruptedExcepti
     }
   }
 
+  public static boolean whetherToSwitchToFetchedCurrentVersion(
+      String storeName,
+      Set<Integer> activeVersions,
+      AtomicInteger currentVersion,
+      int fetchedCurrentVersion,
+      int partitionCountForFetchedCurrentVersion,
+      Map<Integer, List<String>> routingInfo) {
+    if (fetchedCurrentVersion == currentVersion.get()) {
+      // Already same
+      return false;
+    }
+
+    boolean isPartitionResourcesReadyForFetchedCurrentVersion = isPartitionResourcesReady(
+        storeName,
+        fetchedCurrentVersion,
+        partitionCountForFetchedCurrentVersion,
+        routingInfo);
+    boolean isCurrentVersionActive = activeVersions.contains(currentVersion.get());
+    if (!isCurrentVersionActive || isPartitionResourcesReadyForFetchedCurrentVersion) {
+      if (!isCurrentVersionActive && !isPartitionResourcesReadyForFetchedCurrentVersion) {
+        LOGGER.warn(
+            "Current version: {} is not active anymore for store: {}, updating to fetched current version: {}"
+                + " even the partition resources are not ready",
+            currentVersion.get(),
+            storeName,
+            fetchedCurrentVersion);
+      }
+      return true;
+    }
+    // If the partition resources are not ready for the fetched current version, we do not update the current
+    // version to avoid breaking the client.
+    LOGGER.warn(
+        "Partition resources are not ready for store: {}, version: {}. "
+            + "Current version will not be updated to fetched current version: {}, but continue to use version: {}",
+        storeName,
+        fetchedCurrentVersion,
+        fetchedCurrentVersion,
+        currentVersion.get());
+    return false;
+  }
+
+  public static boolean isPartitionResourcesReady(
+      String storeName,
+      int version,
+      int partitionCount,
+      Map<Integer, List<String>> routingInfo) {
+    if (partitionCount <= 0) {
+      LOGGER.warn("Partition count is not valid: {} for store: {}, version: {}", partitionCount, storeName, version);
+      return false;
+    }
+    if (routingInfo == null || routingInfo.isEmpty()) {
+      LOGGER.warn("Routing info is not valid for store: {}, version: {}", storeName, version);
+      return false;
+    }
+    if (routingInfo.size() != partitionCount) {
+      LOGGER.warn(
+          "Routing info size: {} does not match partition count: {} for store: {}, version: {}",
+          routingInfo.size(),
+          partitionCount,
+          storeName,
+          version);
+      return false;
+    }
+    for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+      List<String> replicas = routingInfo.get(partitionId);
+      if (replicas == null || replicas.isEmpty()) {
```

Comment:
I thought we originally wanted to delay the version swap if we don't have sufficient ready to serve replicas as push strategy? i.e. replicas.size() >= 2 (or replication factor - 1).

Reply:
`replicas.size() >= 2` could be too aggressive, and this logic is aligned with the Router behavior:
https://github.com/linkedin/venice/blob/main/services/venice-router/src/main/java/com/linkedin/venice/router/api/VeniceVersionFinder.java#L191
If we want to enforce the minimum replica count to be n - 1, we need to enforce it in the backend since it can carry many risks on client side when some resources have less online replicas because of some node swapping.